### PR TITLE
Add containerSecurityContext

### PR DESCRIPTION
**Pull request by nikonorovi**
Signed-off-by: Igor Nikonorov <igor.nikonorov@syncier.com>

### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

### Checklist

- [x] [DCO](https://github.com/jenkinsci/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] CHANGELOG.md was updated

### CODE REVIEW

File: charts/jenkins/CHANGELOG.md [hunk 1]
```diff
@@ -12,11 +12,14 @@ Use the following links to reference issues, PRs, and commits prior to v2.6.0.
 The change log until v1.5.7 was auto-generated based on git commits.
 Those entries include a reference to the git commit to be able to get more details.
 
+## 3.8.9
+
+Add Added containerSecurityContext
```
Comment (torstenwalter) [inline charts/jenkins/CHANGELOG.md:17]:
```suggestion
Added containerSecurityContext
```

File: charts/jenkins/Chart.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v2
 name: jenkins
 home: https://jenkins.io/
-version: 3.8.8
+version: 3.8.9
```
Comment (torstenwalter) [inline charts/jenkins/Chart.yaml:4]:
```suggestion
version: 3.9.0
```

Let's bump minor version as it's kind of a new feature.

File: charts/jenkins/VALUES_SUMMARY.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ The following tables list the configurable parameters of the Jenkins chart and t
 | `controller.sidecars.configAutoReload.image` | Image which triggers the reload | `kiwigrid/k8s-sidecar:0.1.144`           |
 | `controller.sidecars.configAutoReload.reqRetryConnect` | How many connection-related errors to retry on  | `10`          |
 | `controller.sidecars.configAutoReload.env` | Environment variables for the Jenkins Config as Code auto-reload container  | Not set |
+| `controller.sidecars.configAutoReload.containerSecurityContext` | Enable container security context (enabled if `runAsUser`, `runAsGroup`, `readOnlyRootFilesystem` or `allowPrivilegeEscalation` are set) | Not set |
```
Comment (torstenwalter) [inline charts/jenkins/VALUES_SUMMARY.md:34]:
```suggestion
| `controller.sidecars.configAutoReload.containerSecurityContext` | Enable container security context | Not set |
```

There are no if else statements which enable it automatically.

File: charts/jenkins/values.yaml [hunk 4]
```diff
@@ -116,6 +116,12 @@ controller:
   #   runAsNonRoot: true
   #   supplementalGroups: [1000]
   #   # capabilities: {}
+  # Container securityContext
+  containerSecurityContext:
+    runAsUser: 1000
+    runAsGroup: 1000
```
Comment (torstenwalter) [inline charts/jenkins/values.yaml:122]: These two fields are also defined somewhere else. Maybe we could use `tpl` function when rendering it and reference the other values here.

https://helm.sh/docs/howto/charts_tips_and_tricks/#using-the-tpl-function
https://austindewey.com/2021/02/22/using-the-helm-tpl-function-to-refer-values-in-values-files/

**Comment by torstenwalter** [comment]
```
io.jenkins.tools.pluginmanager.impl.DownloadPluginException: Cannot create a temporary directory for downloads
	at io.jenkins.tools.pluginmanager.impl.PluginManager.downloadPlugins(PluginManager.java:552)
	at io.jenkins.tools.pluginmanager.impl.PluginManager.start(PluginManager.java:225)
	at io.jenkins.tools.pluginmanager.impl.PluginManager.start(PluginManager.java:171)
	at io.jenkins.tools.pluginmanager.cli.Main.main(Main.java:70)
Caused by: java.nio.file.FileSystemException: /tmp/plugin-installation-manager-downloads4023729702115872017: Read-only file system
```

Looks like the readOnlyRootFilesystem does not work for the init container which wants to download plugins.
@timja Do you think that there are other locations where Jenkins nedds write access outside of already mounted volumes?

**Comment by timja** [comment]
> Do you think that there are other locations where Jenkins nedds write access outside of already mounted volumes?

not that I know of, bit risky though...
That code is just writing to the system temporary directory which I would say will need to be writeable.

**Comment by dduportal** [comment]
Do you think that specifying the `/tmp` to mount a tmpfs by default would help?

**Comment by torstenwalter** [comment]
> Do you think that specifying the `/tmp` to mount a tmpfs by default would help?

I think it would.

**Comment by torstenwalter** [comment]
> > Do you think that specifying the `/tmp` to mount a tmpfs by default would help?
> 
> I think it would.

Just not sure if tmpfs is the best option of just a default emptyVolume (not memor backed).

To quote the docs:

> Depending on your environment, emptyDir volumes are stored on whatever medium that backs the node such as disk or SSD, or network storage. However, if you set the emptyDir.medium field to "Memory", Kubernetes mounts a tmpfs (RAM-backed filesystem) for you instead. While tmpfs is very fast, be aware that unlike disks, tmpfs is cleared on node reboot and any files you write **count against your container's memory limit**.

**Comment by timja** [comment]
tmpfs might be ok

**Comment by dduportal** [comment]
> > > Do you think that specifying the `/tmp` to mount a tmpfs by default would help?
> > 
> > 
> > I think it would.
> 
> Just not sure if tmpfs is the best option of just a default emptyVolume (not memor backed).
> 
> To quote the docs:
> 
> > Depending on your environment, emptyDir volumes are stored on whatever medium that backs the node such as disk or SSD, or network storage. However, if you set the emptyDir.medium field to "Memory", Kubernetes mounts a tmpfs (RAM-backed filesystem) for you instead. While tmpfs is very fast, be aware that unlike disks, tmpfs is cleared on node reboot and any files you write **count against your container's memory limit**.

Good point. We should allow the end user to choose memory backed or not (your proposal for emptyDir makes sense).
- The risk with when using a filesystem is related to safety of the content (that should not be available for other pod processes), but it ensures that it always work without risk for the memory quotas

**Comment by torstenwalter** [comment]
empty dir volumes are not available for other pods. They have the same lifecycle as the pod.

File: charts/jenkins/values.yaml (hunk 4 above)
Comment (torstenwalter) [reply charts/jenkins/values.yaml:122]: I double checked. Let's stick to your proposal as the other fields I mentioned are deprecated.

File: charts/jenkins/VALUES_SUMMARY.md [hunk 5]
```diff
@@ -112,6 +113,7 @@ The following tables list the configurable parameters of the Jenkins chart and t
 | `controller.runAsUser`                | Deprecated in favor of `controller.podSecurityContextOverride`.  uid that jenkins runs with. | `1000`                                    |
 | `controller.fsGroup`                  | Deprecated in favor of `controller.podSecurityContextOverride`.  uid that will be used for persistent volume. | `1000`                             |
 | `controller.podSecurityContextOverride` | Completely overwrites the contents of the pod security context, ignoring the values provided for `runAsUser`, and `fsGroup`. | Not set |
+| `controller.containerSecurityContext`    | Allow to control securityContext for the jenkins container. | Not set |
```
Comment (torstenwalter) [inline charts/jenkins/VALUES_SUMMARY.md:116]: Actually it is set.

**Comment by torstenwalter** [comment]
@nikonorovi I checked why it's failing. `ci/other-values.yaml` specified `overwritePluginsFromImage: false`. In that case one actually there is no volume mount. Actually when this is set it does not make sence to download plugins at all as they will not be used by the `jenkins` container.

For this PR I would suggest to set `containerSecurityContext` to an empty value in `ci/other-values.yaml`:

```
controller:
  overwritePluginsFromImage: false
  containerSecurityContext: ""
  runAsUser: 0
  fsGroup: 1000
```

**Review by torstenwalter — APPROVED** [review]
LGTM @nikonorovi thank you for this security improvement!

@timja could you have another look?

**Comment by dduportal** [comment]
If you want a real life case, we (the Jenins infra team) can try this new version; before merging, on our private instance (infra.ci.jenkins.io) which uses the weekly release of Jenkins (e.g. updated at least once per week). Interested?

**Comment by torstenwalter** [comment]
> If you want a real life case, we (the Jenins infra team) can try this new version; before merging, on our private instance (infra.ci.jenkins.io) which uses the weekly release of Jenkins (e.g. updated at least once per week). Interested?

Absolutely. Would love to get that feedback! How quickly could you test it?

**Comment by dduportal** [comment]
> > If you want a real life case, we (the Jenins infra team) can try this new version; before merging, on our private instance (infra.ci.jenkins.io) which uses the weekly release of Jenkins (e.g. updated at least once per week). Interested?
> 
> Absolutely. Would love to get that feedback! How quickly could you test it?

In the upcoming hour :) I'll let you know as soon as it 's applied

**Comment by dduportal** [comment]
@torstenwalter for info, the PR https://github.com/jenkins-infra/charts/pull/1750 is about deploying the current `3.9.0` from this PR.

Before merging, we have the following (important, e.g. outside the label chart/version change) `diff`:

```diff
jenkins-infra, jenkins-infra, StatefulSet (apps) has changed:
...
    labels:
      "app.kubernetes.io/name": 'jenkins'
-     "helm.sh/chart": "jenkins-3.8.9"
+     "helm.sh/chart": "jenkins-3.9.0"
      "app.kubernetes.io/managed-by": "Helm"
      "app.kubernetes.io/instance": "jenkins-infra"
...
            image: "jenkinsciinfra/jenkins-weekly:0.31.0-2.322"
            imagePullPolicy: "Always"
+           securityContext:
+             allowPrivilegeEscalation: false
+             readOnlyRootFilesystem: true
+             runAsGroup: 1000
+             runAsUser: 1000
            command: [ "sh", "/var/jenkins_config/apply_config.sh" ]
            resources:
...
            image: "jenkinsciinfra/jenkins-weekly:0.31.0-2.322"
            imagePullPolicy: "Always"
+           securityContext:
+             allowPrivilegeEscalation: false
+             readOnlyRootFilesystem: true
+             runAsGroup: 1000
+             runAsUser: 1000
            args: [ "--httpPort=8080"]
            env:
...
              - name: jenkins-cache
                mountPath: /var/jenkins_cache
+             - mountPath: /tmp
+               name: tmp-volume
          - name: config-reload
            image: "kiwigrid/k8s-sidecar:1.14.2"
            imagePullPolicy: IfNotPresent
+           securityContext:
+             allowPrivilegeEscalation: false
+             readOnlyRootFilesystem: true
            env:
              - name: POD_NAME
...
            claimName: jenkins-infra
        - name: sc-config-volume
+         emptyDir: {}
+       - name: tmp-volume
          emptyDir: {}
        - name: admin-secret
...
```

important note: we use our own image with its own plugins (source https://github.com/jenkins-infra/docker-jenkins-weekly)

**Comment by dduportal** [comment]
it worked superbly on the (private) instance infra.ci.jenkins.io 👏

**Review by timja — APPROVED** [review]

**Pull request merged**